### PR TITLE
Command line apps use libctx and property query

### LINKED ISSUE #15686: apps are inconsistent about using a library context and property query
When fetching algorithms, the various commands are inconsistent about using the app's library context and property query.

### PR DESCRIPTION
**Pull request by paulidale**
In many places the command line application was not passing it's library context and property query to `XXX_fetch` and `XXX_do_all_provider`.  This addresses this oversight.

Also added a _-digest_ option to the SPKAC command.

Fixes  #15686 & #15683

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/cms.c [hunk 1]
```diff
@@ -721,15 +721,15 @@ int cms_main(int argc, char **argv)
         goto end;
 
     if (digestname != NULL) {
-        if (!opt_md(digestname, &sign_md))
```
Comment (slontis) [inline apps/cms.c:724]: A smarter thing to do would have been to just put the fetch inside opt_md

File: apps/fipsinstall.c [hunk 2]
```diff
@@ -431,7 +431,7 @@ int fipsinstall_main(int argc, char **argv)
     if (read_buffer == NULL)
         goto end;
 
-    mac = EVP_MAC_fetch(NULL, mac_name, NULL);
```
Comment (slontis) [inline apps/fipsinstall.c:434]: Not sure this one is useful..

File: apps/speed.c [hunk 3]
```diff
@@ -2252,13 +2256,14 @@ int speed_main(int argc, char **argv)
     }
 
     if (doit[D_EVP_CMAC]) {
-        EVP_MAC *mac = EVP_MAC_fetch(NULL, "CMAC", NULL);
+        EVP_MAC *mac = EVP_MAC_fetch(app_get0_libctx(), "CMAC",
```
Comment (slontis) [inline apps/speed.c:2259]: This can leak the mac in subsequent gotos

File: apps/cms.c (hunk 1 above)
Comment (paulidale) [reply apps/cms.c]: I got rid of this change.  Too big...

File: apps/speed.c (hunk 3 above)
Comment (paulidale) [reply apps/speed.c:2259]: And it didn't before?????

File: apps/fipsinstall.c (hunk 2 above)
Comment (paulidale) [reply apps/fipsinstall.c:434]: I don't think it changes anything either but it brings all uses into consistency.  We never know where the copy/paste source will be....

File: apps/speed.c (hunk 3 above)
Comment (slontis) [reply apps/speed.c:2259]: You just focused my attention on it :)

Comment (paulidale) [reply apps/speed.c:2259]: You need blinkers :)

Comment (paulidale) [reply apps/speed.c]: I've pushed a fix.

File: test/recipes/20-test_spkac.t [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+plan tests => 4;
+
+# For the tests below we use the cert itself as the TBS file
+
+SKIP: {
+    skip "RSA is not supported by this OpenSSL build", 4
+        if disabled("rsa");
+
+    ok(run(app([ 'openssl', 'spkac', '-key', srctop_file("test", "testrsa.pem"),
```
Comment (slontis) [inline test/recipes/20-test_spkac.t:27]: skip if no md5?

Comment (paulidale) [reply test/recipes/20-test_spkac.t:30]: Done.

File: apps/include/app_libctx.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+#ifndef OSSL_APPS_LIBCTX_H
+# define OSSL_APPS_LIBCTX_H
+
+OSSL_LIB_CTX *app_create_libctx(void);
+OSSL_LIB_CTX *app_get0_libctx(void);
+int app_set_propq(const char *arg);
+const char *app_get0_propq(void);
+
+#endif
```
Comment (levitte) [inline apps/include/app_libctx.h:18]: I real benefit would be if the implementation was moved from `apps/lib/apps.c` to `apps/lib/app_libctx.c` while you're doing this.

Comment (paulidale) [reply apps/include/app_libctx.h:18]: This wouldn't help.  `opt_printf_stderr()` would still be sucked in which pulls in apps.o

I don't remember the specifics of not wanting to `#include "apps.h"` in opt.c beyond knowing there was a reason.

Comment (slontis) [reply apps/include/app_libctx.h:18]: Because it is used in the tests, not just the apps (and apps.h should not get pulled into tests)

Comment (paulidale) [reply apps/include/app_libctx.h:20]: I've moved them.

**Comment by paulidale** [comment]
ping?

File: apps/include/app_libctx.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_APPS_LIBCTX_H
+# define OSSL_APPS_LIBCTX_H
+
+#include <openssl/types.h>
```
Comment (slontis) [inline apps/include/app_libctx.h:13]: # include

Comment (paulidale) [reply apps/include/app_libctx.h:13]: Fixed

File: apps/lib/apps.c [hunk 7]
```diff
@@ -328,50 +326,13 @@ static char *app_get_pass(const char *arg, int keepbio)
...
-}
-
 CONF *app_load_config_bio(BIO *in, const char *filename)
 {
     long errorline = -1;
     CONF *conf;
     int i;
 
-    conf = NCONF_new_ex(app_libctx, NULL);
+    conf = NCONF_new_ex(app_get0_libctx(), NULL);
```
Comment (slontis) [inline apps/lib/apps.c:335]: Hard to know when to use app_get0_propq() ?

Comment (paulidale) [reply apps/lib/apps.c:335]: That argument is a `CONF_METHOD *`, so no.  Still, there are a lot of calls where a propq could be passed and I've not searched for them all.

**Comment by paulidale** [comment]
Ping for review

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the feedback.

**Pull request closed without merging**